src/vm/i386/umthunkstub.S

```diff
@@ -25,8 +26,152 @@ NESTED_ENTRY TheUMEntryPrestub, _TEXT, UnhandledExceptionHandlerUnix
+#define UMThunkStub_LOCALVARS               (2*4)   // UMEntryThunk*, Thread*
+#define UMThunkStub_INT_ARG_SPILL           (2*4)   // for save ecx, edx
+#define UMThunkStub_UMENTRYTHUNK_OFFSET     (UMThunkStub_SAVEDREG+4)
+#define UMThunkStub_THREAD_OFFSET           (UMThunkStub_UMENTRYTHUNK_OFFSET+4)
+#define UMThunkStub_INT_ARG_OFFSET          (UMThunkStub_THREAD_OFFSET+4)
+#define UMThunkStub_FIXEDALLOCSIZE          (UMThunkStub_LOCALVARS+UMThunkStub_INT_ARG_SPILL+4)
+
+// return address                           <-- entry ESP
+// saved ebp                                <-- EBP
+// saved ebx
+// saved esi
+// saved edi
+// UMEntryThunk*
+// Thread*
+// save ecx
+// save edx
+// {optional stack args passed to callee}   <-- new esp
+
+    PROLOG_BEG
+    PROLOG_PUSH ebx
+    PROLOG_PUSH esi
+    PROLOG_PUSH edi
+    PROLOG_END
+    sub     esp, UMThunkStub_FIXEDALLOCSIZE
+
+    mov     dword ptr [ebp - UMThunkStub_INT_ARG_OFFSET], ecx
+    mov     dword ptr [ebp - UMThunkStub_INT_ARG_OFFSET - 0x04], edx
+
+    mov     dword ptr [ebp - UMThunkStub_UMENTRYTHUNK_OFFSET], eax
+
+    call    C_FUNC(GetThread)
+    test    eax, eax
+    jz      LOCAL_LABEL(DoThreadSetup)
+
+LOCAL_LABEL(HaveThread):
+
+    mov     dword ptr [ebp - UMThunkStub_THREAD_OFFSET], eax
+
+    // FailFast if a native callable method invoked via ldftn and calli.
+    cmp     dword ptr [eax + Thread_m_fPreemptiveGCDisabled], 1
+    jz      LOCAL_LABEL(InvalidTransition)
+
+    // disable preemptive GC
+    mov     dword ptr [eax + Thread_m_fPreemptiveGCDisabled], 1
+
+    // catch returning thread here if a GC is in progress
+    PREPARE_EXTERNAL_VAR g_TrapReturningThreads, eax
+    cmp     eax, 0
+    jnz     LOCAL_LABEL(DoTrapReturningThreadsTHROW)
+
+LOCAL_LABEL(InCooperativeMode):
+
+#if _DEBUG
+    mov     eax, dword ptr [ebp - UMThunkStub_THREAD_OFFSET]
+    mov     eax, dword ptr [eax + Thread__m_pDomain]
+    mov     esi, dword ptr [eax + AppDomain__m_dwId]
+
+    mov     eax, dword ptr [ebp - UMThunkStub_UMENTRYTHUNK_OFFSET]
+    mov     edi, dword ptr [eax + UMEntryThunk__m_dwDomainId]
+
+    cmp     esi, edi
+    jne     LOCAL_LABEL(WrongAppDomain)
+#endif
+
+    mov     eax, dword ptr [ebp - UMThunkStub_UMENTRYTHUNK_OFFSET]
+    mov     ebx, dword ptr [eax + UMEntryThunk__m_pUMThunkMarshInfo]
+    mov     eax, dword ptr [ebx + UMThunkMarshInfo__m_cbActualArgSize]
+    test    eax, eax
+    jnz     LOCAL_LABEL(UMThunkStub_CopyStackArgs)
+
+LOCAL_LABEL(UMThunkStub_ArgumentsSetup):
+
+    mov     ecx, dword ptr [ebp - UMThunkStub_INT_ARG_OFFSET]
+    mov     edx, dword ptr [ebp - UMThunkStub_INT_ARG_OFFSET - 0x04]
+
+    mov     eax, dword ptr [ebp - UMThunkStub_UMENTRYTHUNK_OFFSET]
+    mov     ebx, dword ptr [eax + UMEntryThunk__m_pUMThunkMarshInfo]
+    mov     ebx, dword ptr [ebx + UMThunkMarshInfo__m_pILStub]
+
+    call    ebx
+
+LOCAL_LABEL(PostCall):
+
+    mov     eax, dword ptr [ebp - UMThunkStub_THREAD_OFFSET]
```

Comment:
Thank you. I'll fix it. But current code fails before returning from line 114, `call ebx`.